python/cudf/cudf/core/buffer/tenable_buffer.py

```diff
@@ -0,0 +1,302 @@
+        self,
+        base: TenableBuffer,
+        offset: int,
+        size: int,
+        *,
+        passthrough_attributes: Container[str] = ("exposed",),
+    ) -> None:
+        if size < 0:
+            raise ValueError("size cannot be negative")
+        if offset < 0:
+            raise ValueError("offset cannot be negative")
+        if offset + size > base.size:
+            raise ValueError(
+                "offset+size cannot be greater than the size of base"
+            )
+        self._base = base
+        self._offset = offset
+        self._size = size
+        self._owner = base
+        self._passthrough_attributes = passthrough_attributes
+        base._slices.add(self)
+
+    def __getattr__(self, name):
+        if name in self._passthrough_attributes:
+            return getattr(self._base, name)
+        raise AttributeError(
+            f"{self.__class__.__name__} object has no attribute {name}"
+        )
+
+    def _getitem(self, offset: int, size: int) -> BufferSlice:
+        return BufferSlice(
+            base=self._base, offset=offset + self._offset, size=size
+        )
+
+    @property
+    def exposed(self) -> bool:
+        return self._base.exposed
+
+    def get_ptr(self, mode: str = "write") -> int:
+        if mode == "write" and cudf.get_option("copy_on_write"):
+            self.make_single_owner_inplace()
+        return self._base.get_ptr(mode=mode) + self._offset
+
+    def memoryview(
+        self, *, offset: int = 0, size: Optional[int] = None
+    ) -> memoryview:
+        return self._base.memoryview(offset=self._offset + offset, size=size)
+
+    def copy(self, deep: bool = True) -> BufferSlice:
+        """Return a copy of Buffer.
+
+        What actually happens when `deep == False` depends on the
+        "copy_on_write" option. When copy-on-write is enabled, a shallow copy
+        because a deep copy if the buffer has been exposed. This is because we
+        have no control over knowing if the data is being modified when the
+        buffer has been exposed to third-party.
+
+        Parameters
+        ----------
+        deep : bool, default True
+            The semantics when copy-on-write is disabled:
+                - If True, returns a deep copy of the underlying Buffer data.
+                - If False, returns a shallow copy of the Buffer pointing to
+                  the same underlying data.
+            The semantics when copy-on-write is enabled:
+                - Always a deep copy of the underlying Buffer data.
+
+        Returns
+        -------
+        BufferSlice
+            A slice pointing to either a new or the existing base buffer
+            depending on the expose status of the base buffer and the
+            copy-on-write option (see above).
+        """
+        if deep or not cudf.get_option("copy_on_write"):
+            base_copy = self._base.copy(deep=deep)
+        else:
+            base_copy = self._base.copy(deep=self.exposed)
```

Comment:
Maybe rewrite as:
```
if cudf.get_option("copy_on_write"):
    base_copy = self._base.copy(deep=deep or self.exposed)
else:
    base_copy = self._base.copy(deep=deep)
```

Which I _think_ makes it clearer that the exposure status only matters for the CoW case.

Reply:
Agree, fixed